Implement dynamic previous hitobject retention for Skill class

Syriiin opened the pull request:
This PR is part of #11407 which I have split into 4 smaller PRs for the sake of backporting to osu-stable in smaller chunks.
- Add a list of mods to `Skill` class (#11687)
- Implement dynamic previous hitobject retention for `Skill` class (#11688)
- Refactor to encapsulate strain logic into `Skill` class (#11689)
    - Refactor to abstract out strain logic into `StrainSkill` class (#11690, branched from #11689)

# Changes

- Replaces `LimitedCapacityStack` with a new `ReverseQueue` collection
- Adds overridable `RemoveExtraneousHistory` and `AddToHistory` methods for adding and removing objects to and from the queue
- Adds overridable `PreviousCollectionSoftCapacity` for optimising `ReverseQueue` per `Skill` implementation

# Reason

There is no reason we should be limiting skills to knowing only the previous 2 objects. This originally existed as an angle implementation detail of the original pp+ codebase which made its way here, but didn't get used in the same way.
There are valid use cases for dynamic object history retention, such calculating reading density where you might want to retain as many objects as are visible when the current one appears.

Syriiin committed:
Implement dynamic previous hitobject retention for Skill class

There is no reason we should be limiting skills to knowing only the previous 2 objects. This originally existed as an angle implementation detail of the original pp+ codebase which made its way here, but didn't get used in the same way.

smoogipoo commented:
Needs a rebase onto https://github.com/ppy/osu/pull/11690 / master when it's merged.

I'm not sure I really understand the API around `ReverseQueue`. It dequeues the oldest item (so it acts like a normal queue), but then it indexes from the newest item?
Couldn't you just implement this using `Queue`, and wrapping an indexer around it?

smoogipoo commented:
Oh, turns out you can't index a `Queue<T>`... The xmldocs need to better explain how this works though.

smoogipoo commented on osu.Game/Rulesets/Difficulty/Utils/ReverseQueue.cs:
```diff
@@ -0,0 +1,110 @@
+
+namespace osu.Game.Rulesets.Difficulty.Utils
+{
+    /// <summary>
+    /// An indexed queue where items are indexed beginning from the end instead of the start.
+    /// </summary>
```
```suggestion
    /// <summary>
    /// An indexed queue where items are indexed beginning from the most recently enqueued item.
    /// </summary>
```

smoogipoo commented on osu.Game/Rulesets/Difficulty/Utils/ReverseQueue.cs:
```diff
@@ -0,0 +1,110 @@
+        }
+
+        /// <summary>
+        /// Dequeues an item from the <see cref="ReverseQueue{T}"/> and returns it.
```
```suggestion
        /// Dequeues the least recently enqueued item from the <see cref="ReverseQueue{T}"/> and returns it.
```

smoogipoo commented on osu.Game/Rulesets/Difficulty/Utils/ReverseQueue.cs:
```diff
@@ -0,0 +1,110 @@
+        }
+
+        /// <summary>
+        /// Returns an enumerator which enumerates items in the <see cref="ReverseQueue{T}"/> starting from the most recently enqueued one.
```
```suggestion
        /// Returns an enumerator which enumerates items in the <see cref="ReverseQueue{T}"/> starting from the most recently enqueued item.
```

smoogipoo commented on osu.Game/Rulesets/Difficulty/Utils/ReverseQueue.cs:
```diff
@@ -0,0 +1,110 @@
+        /// <summary>
+        /// Returns an enumerator which enumerates items in the <see cref="ReverseQueue{T}"/> starting from the most recently enqueued one.
+        /// </summary>
+        public IEnumerator<T> GetEnumerator()
```
Is this intended to be used a lot? This will incur an allocation every time unless a custom enumerator is added. Something like this: https://github.com/ppy/osu-framework/blob/07cef87ebbb592e97e1b965b3f4147da03570352/osu.Framework/Graphics/Containers/Container.cs#L458-L480

Syriiin replied:
Currently it exists purely to implement the IEnumerable<T> interface, and is only referenced in the test for it.
That said, I do imagine implementations that require a dynamic history will be using the enumerator via foreach loops, so it sounds like a good thing to account for. :+1:

Syriiin replied:
Fixed in bfd3d0cce978c77d4bfa7bbe1a15262e4cc23c80

smoogipoo commented on osu.Game/Rulesets/Difficulty/Skills/Skill.cs:
```diff
@@ -61,19 +68,41 @@ public abstract class Skill
+        /// Remove objects from <see cref="Previous"/> that are no longer needed for calculations from the current object onwards.
+        /// </summary>
+        /// <param name="current">The <see cref="DifficultyHitObject"/> to be processed.</param>
+        protected virtual void RemoveExtraneousHistory(DifficultyHitObject current)
```
I'm not really a fan of this, do you have a situation in which the number of elements in `Previous` changes dynamically? Especially being virtual and all (along with `AddToHistory`).

Syriiin replied:
Yes, the PP+ algorithm I'm porting has a skill that requires time based history retention, instead of a fixed number.

smoogipoo commented on osu.Game/Rulesets/Difficulty/Skills/Skill.cs:
```diff
@@ -61,19 +68,41 @@ public abstract class Skill
         /// </summary>
         public void Process(DifficultyHitObject current)
         {
+            RemoveExtraneousHistory(current);
```
Doesn't this only keep 1 item in the queue at all times (due to `while (Count > 1) Dequeue()`)? Whereas currently it's 2.

Syriiin replied:
Yes, but as it stands, the second object was actually never used in the codebase. It was likely just an artifact of the original calc being based on PP+ which used this for angle calculations.

Syriiin committed:
Update xmldocs

Co-authored-by: Dan Balasescu <smoogipoo@smgi.me>

Syriiin committed:
Add more descriptive xmldoc for ReverseQueue

Syriiin committed:
Merge branch 'master' into diffcalc/dynamic-history

Syriiin committed:
Set default history retention to 0 for Skill and override in StrainSkill

Some skills might not even require history retention, so why waste the allocations?

smoogipoo commented on osu.Game/Rulesets/Difficulty/Skills/Skill.cs:
```diff
@@ -28,12 +36,32 @@ public abstract class Skill
         {
             this.mods = mods;
+
+            if (PreviousCollectionSoftCapacity > 0)
```
Nothing wrong with having a capacity of 1 right? Might as well keep it like that by default and have this always instantiated.

Syriiin replied:
Yea this was just to be super conscious of unnecessary allocations.
Since this only happens once per skill, it's probably not worth the optimisation.
The big part is just leaving the `RemoveExtraneousHistory()` and `AddToHistory()` methods empty.

smoogipoo replied:
Ah right, it'd be empty... Do you have anything against moving this to `StrainSkill` and keeping the implementation private, making with the previous implementation (keep a soft history 1 + private add/remove history methods).

Your PP+ skills could use a local `ReverseQueue` with their own special handling.

smoogipoo commented:
Would the following structure work for you?

```diff
diff --git a/osu.Game/Rulesets/Difficulty/Skills/Skill.cs b/osu.Game/Rulesets/Difficulty/Skills/Skill.cs
index ebd389d823..0b4838718b 100644
--- a/osu.Game/Rulesets/Difficulty/Skills/Skill.cs
+++ b/osu.Game/Rulesets/Difficulty/Skills/Skill.cs
@@ -19,12 +19,9 @@ public abstract class Skill
         protected readonly ReverseQueue<DifficultyHitObject> Previous;
 
         /// <summary>
-        /// Soft capacity of the <see cref="Previous"/> queue.
-        /// <see cref="Previous"/> will automatically resize if it exceeds capacity, but will do so at a very slight performance impact.
-        /// The actual capacity will be set to this value + 1 to allow for storage of the current object before the next can be processed.
-        /// Setting to zero (default) will cause <see cref="Previous"/> to be uninstanciated.
+        /// Number of previous <see cref="DifficultyHitObject"/>s to keep inside the <see cref="Previous"/> queue.
         /// </summary>
-        protected virtual int PreviousCollectionSoftCapacity => 0;
+        protected virtual int HistoryLength => 1;
 
         /// <summary>
         /// Mods for use in skill calculations.
@@ -36,32 +33,17 @@ public abstract class Skill
         protected Skill(Mod[] mods)
         {
             this.mods = mods;
-
-            if (PreviousCollectionSoftCapacity > 0)
-                Previous = new ReverseQueue<DifficultyHitObject>(PreviousCollectionSoftCapacity + 1);
+            Previous = new ReverseQueue<DifficultyHitObject>(HistoryLength + 1);
         }
 
         internal void ProcessInternal(DifficultyHitObject current)
         {
-            RemoveExtraneousHistory(current);
-            Process(current);
-            AddToHistory(current);
-        }
+            while (Previous.Count > HistoryLength)
+                Previous.Dequeue();
 
-        /// <summary>
-        /// Remove objects from <see cref="Previous"/> that are no longer needed for calculations from the current object onwards.
-        /// </summary>
-        /// <param name="current">The <see cref="DifficultyHitObject"/> to be processed.</param>
-        protected virtual void RemoveExtraneousHistory(DifficultyHitObject current)
-        {
-        }
+            Process(current);
 
-        /// <summary>
-        /// Add the current <see cref="DifficultyHitObject"/> to the <see cref="Previous"/> queue (if required).
-        /// </summary>
-        /// <param name="current">The <see cref="DifficultyHitObject"/> that was just processed.</param>
-        protected virtual void AddToHistory(DifficultyHitObject current)
-        {
+            Previous.Enqueue(current);
         }

diff --git a/osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs b/osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs
index 16816be35c..71cee36812 100644
--- a/osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs
+++ b/osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs
@@ -20,8 +20,6 @@ public abstract class StrainSkill : Skill
         /// </summary>
         protected abstract double SkillMultiplier { get; }
 
-        protected override int PreviousCollectionSoftCapacity => 1;
-
         /// <summary>
         /// Determines how quickly strain decays for the given skill.
         /// For example a value of 0.15 indicates that strain decays to 15% of its original value in one second.
@@ -54,17 +52,6 @@ protected StrainSkill(Mod[] mods)
         {
         }
 
-        protected override void RemoveExtraneousHistory(DifficultyHitObject current)
-        {
-            while (Previous.Count > 1)
-                Previous.Dequeue();
-        }
-
-        protected override void AddToHistory(DifficultyHitObject current)
-        {
-            Previous.Enqueue(current);
-        }
-
         /// <summary>
         /// Process a <see cref="DifficultyHitObject"/> and update current strain values accordingly.
         /// </summary>

```

With a sample implementation of something like what I imagine your custom skill being:
```csharp
public class MyCustomSkill : Skill
{
    private readonly ReverseQueue<DifficultyHitObject> previousByTime = new ReverseQueue<DifficultyHitObject>(10);

    public MyCustomSkill(Mod[] mods)
        : base(mods)
    {
    }

    protected override void Process(DifficultyHitObject current)
    {
        while (previousByTime.Count > 0 && previousByTime[previousByTime.Count - 1].StartTime < current.StartTime - 1000)
            previousByTime.Dequeue();

        // Do stuff.

        previousByTime.Enqueue(current);
    }

    public override double DifficultyValue()
    {
        throw new System.NotImplementedException();
    }
}
```
Essentially, having all custom "Previous" logic done on a local queue, but keeping the `Skill` for users that don't have a need for custom logic.

Syriiin commented:
Looks reasonable to me.
The main thing is just having the collection support dynamic sizes.
This does have the caveat of doing a dead code enqueue and dequeue to `Skill.Previous` in the case that that collection isn't used and has `HistoryLength` set to 0.
But if that's alright with you then I see no issue with it.

smoogipoo committed:
Refactor to keep a consistent API

smoogipoo commented on osu.Game/Rulesets/Difficulty/Skills/Skill.cs:
```diff
@@ -16,7 +16,12 @@ public abstract class Skill
         /// <see cref="DifficultyHitObject"/>s that were processed previously. They can affect the strain values of the following objects.
         /// </summary>
-        protected readonly LimitedCapacityStack<DifficultyHitObject> Previous = new LimitedCapacityStack<DifficultyHitObject>(2); // Contained objects not used yet
+        protected readonly ReverseQueue<DifficultyHitObject> Previous;
```
I think it's okay for now? Unsure if interfacing is really worth it - the previous implementation was also open to such manipulation. Calling `Dequeue` isn't going to break the internals of `Skill` as is - only the user's own implementation.

Syriiin replied:
I'm happy either way. I don't think its a likely enough thing to need to protect against, but if we are concerned with API friendliness then its probably not a bad idea since it's trivial to implement a simple interface for the property.

Syriiin commented on osu.Game/Rulesets/Difficulty/Utils/ReverseQueue.cs:
```diff
@@ -0,0 +1,112 @@
+                start = 0;
+            }
+
+            items[(start + Count) % capacity] = item;
```
Would it be preferable to add a comment explaining how the circular array indexing works?
```suggestion
            // Get next index for insertion by taking our starting index, adding the count to find the overflown index, and modding by capacity to wrap around the circular array
            items[(start + Count) % capacity] = item;
```
Or split up the line to self document with descriptive variable names?
```suggestion
            // Get next index for insertion
            var overflowIndex = start + Count;
            var circularIndex = overflowIndex % capacity;
            items[circularIndex] = item;
```

peppy reviewed (dismissed):
blind eye

Syriiin committed:
Implement custom enumerator for ReverseQueue to avoid allocations
